File: plugin/sketchup_nyc_zoning/ui/dialog.js
"use strict";

let currentMode = "borough";

document.querySelectorAll(".tab").forEach(tab => {
  tab.addEventListener("click", () => {
    document.querySelectorAll(".tab").forEach(t => t.classList.remove("active"));
    document.querySelectorAll(".panel").forEach(p => p.classList.add("hidden"));
    tab.classList.add("active");
    currentMode = tab.dataset.mode;
    document.getElementById(`panel-${currentMode}`).classList.remove("hidden");
  });
});

document.getElementById("btn-load").addEventListener("click", () => {
  const params = buildParams();
  if (!params) return;
  setLoading(true);
  sketchup.load_area(JSON.stringify(params));
});

document.getElementById("btn-cancel").addEventListener("click", () => {
  sketchup.cancel_load("{}");
  setStatus("Cancelling...");
});

function buildParams() {
  if (currentMode === "borough") {
    return { mode: "borough", borough: document.getElementById("borough-select").value };
  }
  if (currentMode === "zip") {
    const zip = document.getElementById("zip-input").value.trim();
    if (!/^\d{5}$/.test(zip)) { setStatus("Enter a valid 5-digit ZIP."); return null; }
    return { mode: "zip", zip };
  }
  const fields = ["min-lon","min-lat","max-lon","max-lat"];
  const vals = fields.map(id => parseFloat(document.getElementById(id).value));
  if (vals.some(isNaN)) { setStatus("Fill in all four bbox fields."); return null; }
  return { mode: "bbox", min_lon: vals[0], min_lat: vals[1], max_lon: vals[2], max_lat: vals[3] };
}

function setLoading(on) {
  document.getElementById("btn-load").disabled = on;
  document.getElementById("btn-cancel").disabled = !on;
  if (on) setStatus("Fetching data…");
}

function setStatus(msg) {
  document.getElementById("status").textContent = msg;
}

// Called by Ruby
function onProgress(done, total, pct) {
  document.getElementById("progress-fill").style.width = pct + "%";
  setStatus(`Loading… ${done} / ${total} buildings`);
}

function onLoadComplete(loaded, skipped, cancelled) {
  setLoading(false);
  document.getElementById("progress-fill").style.width = "100%";
  const label = cancelled ? "Cancelled" : "Done";
  setStatus(`${label}: ${loaded} built, ${skipped} skipped.`);
}

function onError(msg) {
  setLoading(false);
  setStatus(`Error: ${msg}`);
}

File: plugin/sketchup_nyc_zoning/ui/dialog.rb
require "json"
require "sketchup.rb"

module NYCZoning
  module Dialog
    DIALOG_OPTIONS = {
      dialog_title:    "NYC Zoning Loader",
      preferences_key: "nyc_zoning_loader",
      style:           UI::HtmlDialog::STYLE_DIALOG,
      resizable:       true,
      width:           420,
      height:          520,
    }.freeze

    HTML_PATH = File.join(NYCZoning::PLUGIN_DIR, "ui", "dialog.html")

    def self.show
      @dialog ||= build_dialog
      @dialog.show
    end

    def self.build_dialog
      dlg = UI::HtmlDialog.new(DIALOG_OPTIONS)
      dlg.set_file(HTML_PATH)

      dlg.add_action_callback("load_area") do |_ctx, params|
        handle_load(JSON.parse(params))
      end

      dlg.add_action_callback("cancel_load") do |_ctx, _params|
        @loader&.cancel!
      end

      dlg
    end
    private_class_method :build_dialog

    def self.handle_load(params)
      bbox = bbox_from_params(params)
      return unless bbox

      client   = McpClient.new
      raw_text = client.call_tool(
        "query_buildings",
        bbox.to_mcp_params.merge(limit: Config["max_features"] || 10_000),
      )
      geojson  = JSON.parse(raw_text)
      features = geojson["features"] || []

      notify_progress(0, features.length)

      @loader = ChunkLoader.new(features, Sketchup.active_model)
      @loader.run do |done, total|
        notify_progress(done, total)
      end

      @dialog.execute_script(
        "onLoadComplete(#{@loader.loaded}, #{@loader.skipped}, #{@loader.cancelled})",
      )
    rescue StandardError => e
      @dialog.execute_script("onError(#{e.message.to_json})")
      NYCZoning::Logger.error("Dialog load error: #{e.message}")
    end

    def self.bbox_from_params(params)
      mode = params["mode"]
      case mode
      when "borough"
        BoroughIndex.bbox(params["borough"])
      when "zip"
        ZipIndex.bbox(params["zip"])
      when "bbox"
        BBox.new(
          params["min_lon"].to_f, params["min_lat"].to_f,
          params["max_lon"].to_f, params["max_lat"].to_f,
        )
      end
    end
    private_class_method :bbox_from_params

    def self.notify_progress(done, total)
      pct = total > 0 ? (done * 100 / total) : 0
      @dialog.execute_script("onProgress(#{done}, #{total}, #{pct})")
    end
    private_class_method :notify_progress
  end
end
